src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -61,19 +72,57 @@ public HttpLoggingMiddleware(RequestDelegate next, IOptionsMonitor<HttpLoggingOp
         /// <returns></returns>HttpResponseLog.cs
         public Task Invoke(HttpContext context)
         {
-            if (!_logger.IsEnabled(LogLevel.Information))
+            var httpEnabled = _httpLogger.IsEnabled(LogLevel.Information);
+            var w3cEnabled = _w3cLogger.IsEnabled(LogLevel.Information);
+            if (!httpEnabled && !w3cEnabled)
             {
                 // Logger isn't enabled.
                 return _next(context);
             }
 
-            return InvokeInternal(context);
+            return InvokeInternal(context, httpEnabled, w3cEnabled);
         }
 
-        private async Task InvokeInternal(HttpContext context)
+        private async Task InvokeInternal(HttpContext context, bool httpEnabled, bool w3cEnabled)
         {
             var options = _options.CurrentValue;
-            RequestBufferingStream? requestBufferingStream = null;
+
+            var w3cList = new List<KeyValuePair<string, object?>>();
+
+            if (w3cEnabled)
+            {
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.DateTime))
+                {
+                    AddToList(w3cList, nameof(DateTime), DateTime.Now.ToString(CultureInfo.InvariantCulture));
```

Comment:
I have a feeling this is supposed to match the date Kestrel writes to the response. I'm guessing this this should be UTC and formatted `DateTimeOffset.UtcNow.ToString("r", CultureInfo.InvariantCulture)` and set after the other middleware has completed, but we should check the IIS W3C logging to confirm.

We should probably try to read that header on the way out and if it's not set, set it yourself using the [DateHeaderValueManager](https://github.com/dotnet/aspnetcore/blob/4dc663a0d46467908f5463d5004b0ae574648f4a/src/Servers/Kestrel/Core/src/Internal/Http/DateHeaderValueManager.cs#L22-L27) which we have to expose somehow to get caching. Even doing some shared source thing where the W3CLoggingMiddleware gets it's own DateHeaderValueManager will be a huge improvement.

Reply:
IIS lets you pick between UTC or local time.

Reply:
with UTC as the default

Reply:
I checked and the [spec](https://www.w3.org/TR/WD-logfile.html) asks for UTC, but the format looked completely different than the `Date: ` header, so it seems I was wrong about the format. It looks like date and time are supposed to be separate fields.

I found https://stackoverflow.com/a/19825775/719967 which is also interesting. Apparently, IIS takes the timestamp when the log is emitted, so I don't think there's any expectation it match the Date header so ignore everything I said about that. That said, caching the results of these ToString() operations should be done like in DateHeaderValueManager, but that could be a follow up issue we make up-for-grabs or something.

Reply:
Marking as resolved